[Console] Modify console output and print multiple modifyable sections

pierredup opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 4.1
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | TBD
| Fixed tickets | 
| License       | MIT
| Doc PR        | symfony/symfony-docs#9304

Add support to create different output sections for the console output.
Each section is it's own 'stream' of output, where the output can be modified (even if there are other output after it). This allows you to modify previous output in the console, either by appending new lines, modifying previous lines or clearing the output. Modifying a sections output doesn't affect the output after that or in other sections.

Some examples of what can be done:

**Overwriting content in a previous section:**

Code:

```php
$section1 = $output->section();
$section2 = $output->section();

$section1->writeln("<comment>Doing something</comment>\n");
usleep(500000);
$section2->writeln('<info>Result of first operation</info>');
usleep(500000);

$section1->overwrite("<comment>Doing something else</comment>\n");
usleep(500000);
$section2->writeln('<info>Result of second operation</info>');
usleep(500000);

$section1->overwrite("<comment>Finishing</comment>\n");
usleep(500000);
$section2->writeln('<info>Last Result</info>');
```

Result:
![overwrite-append](https://user-images.githubusercontent.com/144858/30975030-769f2c46-a471-11e7-819f-c3698b43f0af.gif)

**Multiple Progress Bars:**

Code:

```php
$section1 = $output->section();
$section2 = $output->section();

$progress = new ProgressBar($section1);
$progress2 = new ProgressBar($section2);

$progress->start(100);
$progress2->start(100);

$c = 0;
while (++$c < 100) {
    $progress->advance();

    if ($c % 2 === 0) {
        $progress2->advance(4);
    }

    usleep(500000);
}
```

Result:
![multiple-progress](https://user-images.githubusercontent.com/144858/30975119-b63222be-a471-11e7-89aa-a555cdf3d2e0.gif)

**Modifying content of a table & updating a progress bar:**

Code:

```php
$section1 = $output->section();
$section2 = $output->section();

$progress = new ProgressBar($section1);
$table = new Table($section2);

$table->addRow(['Row 1']);
$table->render();

$progress->start(5);

$c = 0;
while (++$c < 5) {
    $table->appendRow(['Row '.($c + 1)]);

    $progress->advance();

    usleep(500000);
}

$progress->finish();
$section1->clear();
```

Result:
![progress-table](https://user-images.githubusercontent.com/144858/30975176-e332499c-a471-11e7-9d4f-f58b464a53c2.gif)

**Example with Symfony Installer:***

Before:
![sf-installer-old](https://user-images.githubusercontent.com/144858/30975291-40f22106-a472-11e7-8836-bc39139c2d30.gif)

After:
![sf-installer](https://user-images.githubusercontent.com/144858/30975302-4a00acf4-a472-11e7-83ba-88ea9d0f0f3f.gif)

TODO:
- [x] Add unit tests

pierredup committed:
Add section output to console

javiereguiluz commented:
I love this proposal! Just the other day I was playing with Yeoman Generator and they use a similar technique and it amazed me. They make things appear and disappear and change the previous contents, for example when asking several consecutive questions:

![yeoman-generator](https://user-images.githubusercontent.com/73419/30981932-2ddbf9e8-a486-11e7-82c1-f827d233b685.gif)

![yeoman-generator-2](https://user-images.githubusercontent.com/73419/30981937-3072180e-a486-11e7-9236-6a9736feb2cd.gif)

stof commented on src/Symfony/Component/Console/Output/OutputSectionReference.php:
```diff
@@ -0,0 +1,57 @@
+        $buffer = 0;
+
+        foreach ($this->ref as $outputStream) {
+            if (spl_object_id($outputStream) === spl_object_id($output)) {
```
just compare the objects themselves inside of comparing the ids explicitly. It will be easier to understand (and will avoid the dependency on the polyfill).

stof commented on src/Symfony/Component/Console/Output/OutputSectionReference.php:
```diff
@@ -0,0 +1,57 @@
+        $content = '';
+
+        foreach ($this->ref as $outputStream) {
+            if (spl_object_id($outputStream) === spl_object_id($output)) {
```
same here

stof commented:
this is clearly lacking tests.

pierredup commented:
@stof The todo at the end of the description says that tests still needs to be added

yceruto commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,140 @@
+     * @param resource                 $stream
+     * @param OutputSectionReference   $sectionReference
+     * @param int                      $verbosity
+     * @param null                     $decorated
```
`bool|null`? null for auto-guessing

yceruto commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,140 @@
+     * @param OutputSectionReference   $sectionReference
+     * @param int                      $verbosity
+     * @param null                     $decorated
+     * @param OutputFormatterInterface $formatter
```
`OutputFormatterInterface|null`? null to use default OutputFormatter

pierredup replied:
See reply below, we want to use the values from the original console output

yceruto commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,140 @@
+     * @param null                     $decorated
+     * @param OutputFormatterInterface $formatter
+     */
+    public function __construct($stream, OutputSectionReference $sectionReference, $verbosity, $decorated, OutputFormatterInterface $formatter)
```
`..., $verbosity = self::VERBOSITY_NORMAL, $decorated = null, OutputFormatterInterface $formatter = null`?

pierredup replied:
I'm not sure if these values should to be defaulted, as we want to inherit all the values from origin console output

yceruto replied:
It's pragmatic right now, out-the-box we would want to use defaults.

pierredup replied:
This class should never be used on it's own, it's just an implementation detail and should always be constructed from the `ConsoleOutput`, which means it should always inherit the values set there (default or otherwise)

yceruto replied:
That means the PHPDoc of the constructor could be removed altogether if so?

pierredup replied:
Dockblock removed

yceruto reviewed (commented):
Nice feature, thanks!

yceruto commented on src/Symfony/Component/Console/Output/OutputSectionReference.php:
```diff
@@ -0,0 +1,57 @@
+ *
+ * @author Pierre du Plessis <pdples@gmail.com>
+ */
+final class OutputSectionReference
```
I wondering if we can remove the `final` constraint and add `@final` instead?

derrabus replied:
The annotation would only be useful if we wanted to declare the class final after a stable release. Since it's a new class, the `final` keyword would be the way to go, imho.

pierredup replied:
@ostrolocky This is a purely internal class, whether it's final or not should not make any difference to the user as they would never know about this class

joelwurtz commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,140 @@
+        $prevLines = explode("\n", $messages);
+
+        for ($i = 0; $i < count($prevLines); ++$i) {
+            $total += ceil(strlen($prevLines[$i]) / $this->terminal->getWidth()) ?: 1;
```
Should use a better function than strlen (only visible characters must be counted)

chalasr replied:
[Helper::strlen()](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Console/Helper/Helper.php#L52) should be used

ostrolucky commented:
Only first example works properly for me. It also works differently if some part of section is outside of viewport

![2](https://user-images.githubusercontent.com/496233/31050704-322e8e7e-a653-11e7-8a82-3baaa3bb6983.gif)

![1](https://user-images.githubusercontent.com/496233/31050703-2f2f6374-a653-11e7-8c38-6d35888886cb.gif)

pierredup commented:
@ostrolucky Can you please try the examples again, there was some missing changes that I forgot to commit

> It also works based on a viewport

A limitation with the console is that you can only manipulate anything in the viewport. So if any content stretches outside the viewport, it becomes extremely difficult (if not impossible) to modify the contents.

chalasr commented on src/Symfony/Component/Console/Output/ConsoleOutput.php:
```diff
@@ -51,6 +53,18 @@ public function __construct($verbosity = self::VERBOSITY_NORMAL, $decorated = nu
+     */
+    public function section()
+    {
+        return new ConsoleSectionOutput($this->getStream(), $this->outputReference, $this->getVerbosity(), $this->isDecorated(), $this->getFormatter());
```
This method should throw if `!$this->isDecorated()`

pierredup replied:
@chalasr I don't think this should throw as it might cause issues in E.G a CI environment where the output should just continue printing without decoration. If this throws an exception, the commands would fail and not continue to run, which might not be the expected behaviour. I have instead added checks for decoration to skip the ANSI codes

joelwurtz commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,137 @@
+        $prevLines = explode("\n", $messages);
+
+        for ($i = 0; $i < count($prevLines); ++$i) {
+            $total += ceil(Helper::strlen($prevLines[$i]) / $this->terminal->getWidth()) ?: 1;
```
I don't think it's stille the good way to calculate the length of a line in a console, as an example a tabulation must be count as 8 when displaying (or other number depending on the term configuration) or a colore string / background will have special chars which here will be counted, but they must not.

I was confronted with the same behavior and use this implementation : https://github.com/jolicode/asynit/blob/master/src/Output/TestOutput.php#L110 far from being perfect but gives a good start.

ostrolucky commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -279,6 +280,11 @@ public function setRow($column, array $row)
     public function render()
     {
+        if ($this->output instanceof ConsoleSectionOutput) {
+            // TODO: Calculate the number of lines to remove for the table
```
outdated todo?

pierredup replied:
No this still needs to be done, currently it's only calculating the number of rows and headers, but not taking the row separator into account

ostrolucky commented:
This kinda changes behaviour or Table helper. Not sure this is acceptable. Table helper without this patch doesn't redraw it's content. Calling
```php
$table->render();
$table->render();
```

would normally output table twice, but this patch makes it erase first one and replace it with second one

pierredup commented:
The change in behaviour for the table helper is intented. The table is rendered in a section output, that can be modified. This adds a feature to the table helper where you can modify and re-render the table without rendering it twice. If you want the normal behaviour, just pass the normal console output instance to the table and not a sectioned output.

The newline at the end is currently a requirement, as it becomes difficult to navigate the console and change the contents without it. But I am working on a patch.

I'm not sure if we should cater for output woth escape sequences. If a user wants to use escape sequences, they should just use the standard console ouput. If they want to modify the output, then they should use the section output.

This PR is not finished, there are a lot of edge cases and some missing functionality that I'm working on to solve, which I will pusb once it's done. Thanks for the feedback so far

chalasr commented:
@ostrolucky If the Table behavior is different only when using the new output class then that means Table just leverages the new feature. If passing the existing outputs gives the same result as before (reading the code I believe it does) then there is no issue.

ostrolucky commented:
@chalasr I still don't think this feature should affect behaviour of current components. Purpose of this feature is just ability to split screen to sections. How is forcing table to redraw itself related to that? Here is better example why is this wrong:
```php
$table1->render();
$table2->render();
```
You would expect both tables to be shown on the screen. Instead, content of second table will overwrite content of first table, unless explicitly assigned different output section. I don't think this is expected. If you do this in a loop or have branching logic which determines if, when and which table is supposed to be outputted, you will quickly find out that this approach is unmaintainable.

Rest of my points are not so important, solving them would require to make implementation a lot more complicated, I was just not sure if they are important for you so wanted to brought them up. If they aren't, they aren't for me neither.

pierredup commented:
@ostrolucky How will the second table overwrite the first one? If you use different output sections, they will still render both, as the table is only redrawn within it's own section. If you use the normal console output and not the new class, both tables will still be shown as it won't use the new functionality. So IMO there is no unexpected behaviour. The only missing part in this instance is that the new class doesn't take into account the output of the normal console. But that is one of the fixes I'm working on. So if you combine thr normal console output and the output of the new class, then you might run into some rendering issue, but that should be fixed soon

ostrolucky commented:
You didn't continue to read. I did mention that you if you want to avoid this, you must use different output sections and that's complicated when you have any kind of logic. For example:

```php
$section1 = $output->section();
$section2 = $output->section();

$section2->writeln('some stuff');

while (true) {
    $table = new Table($section1);
    $table->setHeaders(['Stuff']);
    $table->addRow([$rand]);
    $table->render();
}
```

Now, how do I output contents of these tables above section2 without redrawing them? Now I am forced to manipulate with order of the sections container directly, because I have to create new output sections on the fly and push them in between.

pierredup commented:
Then just use ` $table = new Table($output);` as you would normally have instead of using `$output->section()`

ostrolucky commented:
That would render table below $section2, wouldn't it?

pierredup commented:
Yes it would. But even with your first example if you want to use a section, the second table wom't overwrite the first one. Each table will still be rendered below each other as you expect it to do, as the table helper only clears the table it rendered and not the entire section output. There is still a todo thay needs to be addressed in the table helper and I'm still busy tweaking the functionality, so I don't think this is an issue

chalasr commented:
I think I agree with @ostrolucky, having `Table` clearing the section seems wrong. 
If one needs to write two consecutive tables in a section so that the last table replaces the first, then one should just `clear()` the output section used to write the first table before writing the last one. 
I mean:

```php
// before
$table = new Table($section)
$table->render();
$table->render();

//after
$table = new Table($section);
$table->render(); // first
$section->clear();
$table->render(); // last
```

doesn't that work already?

pierredup commented:
Yes that does work already, but the being able to append rows to a table seems like a nice feature to me. In most cases I've seen, the data needs to be collected in arrays and added to the table which is then rendered at the end, where it would be easier to just append the rows in the table. What about a new method on the table helper called `append(array $rows)` which clears and re-draws the table?

chalasr commented:
I agree it is nice. I think it's not obvious that `render()` should behave like that because the new output is used to construct the Table, `Table::appendRow()` (to be consistent with the current api) seems better to me because more explicit. I would make it throw if the current output is not a section (or create one internally if an instance of ConsoleOutputInterface?).

ostrolucky commented:
@pierredup How is this going? Do you need help with anything?

Also note that this patch is currently broken, as $content and $lines must be public.

pierredup commented:
@ostrolucky I'm working on it as I get time, but since it can only go into Symfony 4.1 (to be released May next year) there's no rush. I'll have it ready for review after 4.0 is tagged so that I can rebase against master at that point

keradus commented on src/Symfony/Component/Console/Output/ConsoleOutput.php:
```diff
@@ -34,6 +34,8 @@ class ConsoleOutput extends StreamOutput implements ConsoleOutputInterface
      */
     private $stderr;
 
+    private $consoleSectionOutputs = array();
```
typing is missing

keradus commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+ * @author Pierre du Plessis <pdples@gmail.com>
+ * @author Gabriel Ostrolucký <gabriel.ostrolucky@gmail.com>
+ */
+class ConsoleSectionOutput extends StreamOutput
```
does it makes sense to use this class totally without `ConsoleOutput` ?
especially when having the connection between them via `&$sections`

ostrolucky replied:
It's fine to construct this class completely without `ConsoleOutput`. Coupling it to `ConsoleOutput` more strictly does not make sense

pierredup replied:
I don't think this class should ever be used outside of ConsoleOutput, as it is an implementation detail of the console. I agree this class can be marked as internal

chalasr replied:
Well, using this feature as an end user involves to use the public api of this class, marking it as internal without providing an interface would be weird. I'm not sure we want to introduce an interface for that, so letting it as is (maybe `@final`) with some documentation is probably fine.

ostrolucky replied:
As I have shown in different comment, there are situations when its a must to manipulate with order of sections manually, which can be done only when having access to array container. section() method provides only the most basic operation (creating new section at the bottom of the screen)

keradus replied:
That's the thing. manually creating a section without having it anchored to full output class is weird and imo shall not be allowed.
Currently Output is aware about sections and sections one about another. so we have big part which could contain small parts.
Even if it technically possible to create standalone section instance, what is the usage of that?

Thus section concrete class shall IMO be final, created only by output class itself, so interface to provide info how to use it.

ostrolucky replied:
I don't see anything weird about that. ConsoleOutput currently serves only as a simple stupid helper for creating new output sections at the bottom of the screen. I'm out of luck if I need to create another new section at the top of existing sections. Anyway I could currently do this even if that class is final, because as I said ConsoleOutput serves as a stupid helper only and I can bypass it, final or not. So I don't see what does it solve. 

I would need to ask symfony community to stop making final classes though, you cannot anticipate what will user space need to do. `@final` annotation is enough to mark extensions as not covered by BC policy. I wouldn't even do that though, this class has lean API and high potential to be further extended.

Marking it `@internal` is out of the question as @chalasr explained.

keradus replied:
I disagree about all `final`-about statements. But I guess I'm just defensive.

No, he didn't ;) he said it would be weird when done `without providing an interface`

If you construct section out of the box without `ConsoleOutput` helping here, you need to manually interact with other sections or have that `&$sections`. Yet, it's private in both, `SectionOutput` and `ConsoleOutput`. It will be way cleaner if all will be done via a helper, if you need to inject a section (put it not at the end) after some of them were already created, adjust the helper to allows that.
I'm big critic of passing reference around different instances and manually managing them, so if we need to have them, let us make it internal implementation detail, not sth that user need to deal if he try to not use helper.

keradus commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+    {
+        parent::__construct($stream, $verbosity, $decorated, $formatter);
+        array_unshift($sections, $this);
+        $this->sections = &$sections;
```
why ref here?

ostrolucky replied:
It's a must so property is updated when different section appends itself into array

ostrolucky commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+ * @author Pierre du Plessis <pdples@gmail.com>
+ * @author Gabriel Ostrolucký <gabriel.ostrolucky@gmail.com>
+ */
+class ConsoleSectionOutput extends StreamOutput
```
What would be advantage of that when tradeoff is incompatible API? I would like to continue depending on OutputInterface with no changes. Client doesn't need to be aware that its output is going to appear in some section.

keradus replied:
agree with @ostrolucky here
let say I want to make app with 2 sections (top, bottom), and then I sent given sections as an input for some subcommands. They don't need to know if they run on full output or just section of it.

keradus commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -279,6 +280,11 @@ public function setRow($column, array $row)
     {
+        if ($this->output instanceof ConsoleSectionOutput) {
+            // TODO: Calculate the number of lines to remove for the table
+            $this->output->clear(count($this->rows) + count($this->headers) + 1);
```
in one place it doesn't matter. I would rather advice to make a feature for Fixer and apply rule globally on repo, not sometimes

chalasr commented on src/Symfony/Component/Console/Output/ConsoleOutput.php:
```diff
@@ -53,6 +55,16 @@ public function __construct($verbosity = self::VERBOSITY_NORMAL, $decorated = nu
+     *
+     * @return ConsoleSectionOutput
+     */
+    public function section()
```
Return type needed once rebased

chalasr commented on src/Symfony/Component/Console/Output/ConsoleOutputInterface.php:
```diff
@@ -32,4 +32,11 @@ public function getErrorOutput();
+     *
+     * @return ConsoleSectionOutput
+     */
+    // public function section(); // TODO Add this method to the interface in Symfony 4
```
Let's use an `@method` annotation instead, if we really want it to be part of the interface.

chalasr commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @param int $lines Number of lines to clear. If null, then the entire output of this section is cleared
+     */
+    public function clear($lines = null)
```
`int $lines = null`

chalasr commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+     *
+     * @return string
+     */
+    private function popStreamContentUntilCurrentSection($numberOfLinesToClearFromCurrentSection = 0)
```
`: string`

chalasr commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,125 @@
+        return implode('', array_reverse($erasedContent));
+    }
+
+    private function getDisplayLength($text)
```
`(string $text): int`

chalasr reviewed (commented):
Could you rebase and fix the remaining todo?

keradus commented on src/Symfony/Component/Console/Output/ConsoleOutput.php:
```diff
@@ -31,6 +31,8 @@ class ConsoleOutput extends StreamOutput implements ConsoleOutputInterface
 {
     private $stderr;
 
+    private $consoleSectionOutputs = array();
```
please document the content of that array

keradus commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,150 @@
+        return $col;
+    }
+
+    private function getDisplayLength($text): string
```
what is the type of input ?

pierredup commented:
Unit tests added, remaining todos addressed and comments  addressed.
I think this one is now ready

Status: Needs Review

keradus commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,129 @@
+
+    private $terminal;
+
+    public function __construct($stream, array &$sections, $verbosity, $decorated, OutputFormatterInterface $formatter)
```
please add types of input. either in prototype or in phpdoc

keradus reviewed (changes requested):
`ConsoleSectionOutput` is tightened very strong with `ConsoleOutput` as they are sharing internals via reference.
Please disallow creating `ConsoleSectionOutput` on ones own (eg via marking the class as internal and exposing the interface to end user) and keep the possibility to do it only via `ConsoleOutput::section()` you created.

ostrolucky commented:
@keradus  No, stop it. We already talked about this. This class cannot be marked internal unless public interface is provided, which has lot more disadvantages than advantages. Not having ability to construct it also prevents advanced cases like I demonstrated.

pierredup commented:
@keradus I removed the coupling from `ConsoleOutput` and got rid of the reference and used a static property instead. This makes the class usable on it's own (If necessary I'll still add an interface)

ostrolucky reviewed (changes requested):
This introduces global state which cannot be destroyed https://3v4l.org/PKGCL

I think Fabien did plenty of work in past to get rid of such thing.

It also prevents advanced use cases, like I'm saying whole time. For example, I can append new section only below last one. Give me write access to `$sections` outside the class.

pierredup commented:
Inserting sections between others would be a difficult task, even if you have public access to the array. How would you determine where to add a new section in the array? Adding to the beginning or end would be easy, anything else needs some extra boilerplate code and re-indexing the array to prevent overwriting the section. I'm 👎  for allowing access to the sections array as it can cause more trouble than the benefit it can provide. An alternative would be to have named sections, which would make it easier to manage (clearing a section by name, removing a section from the output etc) but that is adding a bit more complexity.

The global state of the static property is needed if the class needs to be used standalone, so that each new section can always be aware of the previous sections, regardless of where the class was instantiated. The only alternative I see is following @keradus advice and make the class internal and expose only an interface, which means the it can't be used as a standalone class and must be tied to the ConsoleOutput class

ostrolucky commented:
> How would you determine where to add a new section in the array? Adding to the beginning or end would be easy, anything else needs some extra boilerplate code and re-indexing the array to prevent overwriting the section

So? That's userspace job, you don't need to worry about that. I will know how I want to rearrange that in my application. I'm not asking to simplify this job, I'm asking to make it possible. At least make it protected, or create getter/setter, so I won't have to use dirty hacks like reflection.

I think you don't need to follow mine and @keradus conflicting advices though yet, let proper symfony members review this and say their opinion.

I just wanted to express myself that I really don't like such constrained code, I presented use case this doesn't allow. I also don't mind having container for sections outside the class at all, unlike @keradus . I think this was used just as an excuse to make this class as restricted as possible, which I'm not fan of.

chalasr commented:
Could making `ConsoleOutputSection$sections` static leads to weird behaviors when running two commands in the same process?

keradus commented:
If one would one to use Sections he will hopefully as well use them in tests, having `ConsoleSectionOutput::$sections` static will make the state consistent across executed tests (=global) and for heavy tests will create constant memory increase (slightly, yet without way to clean it).
For that, :+1: for not having it static.

pierredup commented:
I reverted the changes to make the sections static. So the `ConsoleSectionOutoput` class stays coupled to `ConsoleOutput` and makes it just an implementation detail of the console output and not usable on it's own (Which was the original intention).

I've also removed the changes to the Table helper so that it doesn't change any current behaviour if you use a section output. Instead I have added an `appendRow` method which will add a row the table and redraw the table

ostrolucky reviewed (changes requested):
There is a bug with ProgressBar.

```php
$section1 = $output->section();
$section2 = $output->section();
$section2->writeln('Am I at fault? :(');

$progress = new ProgressBar($section2);
$progress->start(100);

$c = 0;
while (++$c < 100) {
	$progress->advance();

	if ($c % 2 === 0) {
		$section1->writeln("Why are you erasing me in step $c? :(");
	}

	usleep(50000);
}
```

If you notice, ProgressBar uses different section than writeln inside loop, but it affects it anyway.

pierredup commented:
Thanks for picking it up @ostrolucky, it should now be fixed

fabpot commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -252,6 +255,23 @@ public function addRow($row)
+    public function appendRow($row)
+    {
+        if (!$this->output instanceof ConsoleSectionOutput) {
+            throw new \RuntimeException(sprintf('Output should be an instance of %s when calling %s', ConsoleSectionOutput::class, __METHOD__));
```
Missing dot at the end of the message. You should also add `"` around `%s`: `Output should be an instance of "%s" when calling "%s".`

fabpot commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -252,6 +255,23 @@ public function addRow($row)
+        }
+
+        $this->addRow($row);
+
```
blank line can be removed

fabpot commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -311,6 +331,8 @@ public function render()
         $this->renderRowSeparator();
 
         $this->cleanup();
+
```
can be removed

fabpot commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -445,6 +467,20 @@ private function buildTableRows($rows)
+    {
+        $numberOfRows = count($this->rows);
+
+        if (!empty($this->headers)) {
```
I would avoid using `empty`.

fabpot commented on src/Symfony/Component/Console/Output/ConsoleOutputInterface.php:
```diff
@@ -27,4 +29,11 @@
+     *
+     * @return ConsoleSectionOutput
+     */
+    // public function section(); // TODO Add this method to the interface in Symfony 5
```
I would not add this `section()` method in this interface as this is not really a required feature. If needed, we could create an additional interface, not sure if it's worth it though.

fabpot commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,143 @@
+     * @param ConsoleSectionOutput[]   $sections
+     * @param int                      $verbosity
+     * @param bool                     $decorated
+     * @param OutputFormatterInterface $formatter
```
The last 3 `@param` can be removed.

fabpot commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @param int $numberOfLinesToClearFromCurrentSection
+     *
+     * @return string
```
Both `@return` and `@param` can be removed. As a rule of thumb, we don't add phpdocs that do not add more information than already contained in the code itself.

ostrolucky commented:
There is at least one relevant test failure https://travis-ci.org/symfony/symfony/jobs/344188508#L4298

ostrolucky commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -445,6 +465,20 @@ private function buildTableRows($rows)
+
+        ++$numberOfRows; // Add row for footer separator
+
+        return $numberOfRows;
```
return count($this->rows) + count($this->headers) + 2 + 1;

ostrolucky reviewed (changes requested):
Here's one more bug with section overwriting content of other section:

```php
$section1 = $output->section();
$section2 = $output->section();

$progress = new ProgressBar($section1);
$progress2 = new ProgressBar($section2);
ProgressBar::setFormatDefinition('test', '[%bar%] Fruitcake marzipan toffee. Cupcake gummi bears tart dessert ice cream chupa chups cupcake chocolate bar sesame snaps. Croissant halvah cookie jujubes powder macaroon. Fruitcake bear claw bonbon jelly beans oat cake pie muffin');
$progress2->setFormat('test');

$progress->start(100);
$progress2->start(100);

$c = 0;
while (++$c < 100) {
	$progress->advance();

	if ($c % 2 === 0) {
		$progress2->advance(4);
	}

	usleep(500000);
}
```


Also, please get rid of that one `empty` left, neither fabien and neither me like those :upside_down_face:

ostrolucky commented:
@pierredup Feature freeze is coming. I would like to ensure this is ready soon. Are you working on resolving this, or should I take a look?

pierredup commented:
Mentioned bug fixed.

Status: Needs Review

ostrolucky commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -252,6 +255,22 @@ public function addRow($row)
         return $this;
     }
 
+    public function appendRow($row)
```
How come no public methods in this class have any PHPDoc? This is gonna get confusing when differentiating between `addRow` and `appendRow`

ostrolucky reviewed (changes requested):
I'm really happy with ProgressBar and normal writeln parts of this implementation. I already heavily use this patch in [my project](https://github.com/ostrolucky/stdinho), that's why I'm so invested in this and found so many bugs. However, I don't use tables there.

Before approving, I tried all of the examples again and found last example does not work correctly. I know that sample is outdated, but it doesn't work correctly even if I remove ->render calls and replace addRow calls with appendRows. Am I doing something wrong?

```php
$output = new ConsoleOutput();
$section1 = $output->section();
$section2 = $output->section();

$progress = new ProgressBar($section1);
$table = new Table($section2);
$table->appendRow(['Row 1']);
$progress->start(5);
$table->appendRow(['Row 2']);
$progress->advance();

chalasr reviewed (commented):
2 minors comments, looking for @ostrolucky's comment to be addressed before approving

pierredup commented:
Last comments address and table bug fixed. I had to make a change in the way the number of lines to clear is calculated. I tested all the original examples and every one works now as expected.

@ostrolucky Thank you for all the effort in properly reviewing and testing this.

Status: Needs Review

chalasr reviewed (approved):
A new entry in the console CHANGELOG and 👍  for me, great work @pierredup!

pierredup committed:
Added Changelog entry

fabpot commented on src/Symfony/Component/Console/CHANGELOG.md:
```diff
@@ -5,6 +5,7 @@ CHANGELOG
 -----
 
  * added option to run suggested command if command is not found and only 1 alternative is available
+ * added option to modify console output and print multiple modifyable sections
```
modifiable

fabpot commented on src/Symfony/Component/Console/Helper/Table.php:
```diff
@@ -252,6 +256,25 @@ public function addRow($row)
     }
 
+    /**
+     * Add a row to the table, and re-render the table.
```
Adds
re-renders

fabpot commented on src/Symfony/Component/Console/Output/ConsoleSectionOutput.php:
```diff
@@ -0,0 +1,133 @@
+        }
+
+        if ($numberOfLinesToClear > 0) {
+            // Move cursor up n lines
```
move

pierredup committed:
Fix typos

fabpot commented:
Thank you @pierredup.
